src/common/webrequest_curl.cpp

```diff
@@ -362,217 +510,1194 @@ void wxWebAuthChallengeCURL::SetCredentials(const wxWebCredentials& cred)
+{
+    SocketDataMap::iterator it = m_socketData.find(sock);
+    wxEventLoopSourceHandler* srcHandler = NULL;
 
-    // Signal the worker thread to resume work
-    wxMutexLocker lock(m_mutex);
-    m_condition.Signal();
+    if ( it != m_socketData.end() )
+    {
+        // If this socket is already being polled, reuse the old handler. Also
+        // delete the old source object to stop the old polling operations.
+        wxEventLoopSource* oldSrc = it->second;
+        srcHandler = oldSrc->GetHandler();
 
-    return true;
+        delete oldSrc;
+    }
+    else
+    {
+        // Otherwise create a new source handler.
+        srcHandler =
+            new SocketPollerSourceHandler(sock, m_handler);
+    }
+
+    // Get a new source object for these polling checks.
+    bool socketIsPolled = true;
+    int eventSourceFlag = SocketPoller2EventSource(pollAction);
+    wxEventLoopSource* newSrc =
+        wxEventLoopBase::AddSourceForFD(sock, srcHandler, eventSourceFlag);
+
+    if ( newSrc == NULL )
+    {
+        // We were not able to add a source for this socket.
+        wxLogDebug(wxString::Format(
+                       "Unable to create event loop source for %d",
+                       static_cast<int>(sock)));
+
+        delete srcHandler;
+        socketIsPolled = false;
+
+        if ( it != m_socketData.end() )
+        {
+            m_socketData.erase(it);
+        }
+    }
+    else
+    {
+        m_socketData[sock] = newSrc;
+    }
+
+    return socketIsPolled;
 }
 
-void wxWebSessionCURL::CancelRequest(wxWebRequestCURL* request)
+void SourceSocketPoller::StopPolling(wxSOCKET_T sock)
 {
-    // Add the request to a list of threads that will be removed from the curl
-    // multi handle in the worker thread
-    wxCriticalSectionLocker lock(m_cancelled.cs);
-    request->IncRef();
-    m_cancelled.requests.push_back(wxObjectDataPtr<wxWebRequestCURL>(request));
+    SocketDataMap::iterator it = m_socketData.find(sock);
+
+    if ( it != m_socketData.end() )
+    {
+        CleanUpSocketSource(it->second);
+        m_socketData.erase(it);
+    }
 }
 
-wxVersionInfo  wxWebSessionCURL::GetLibraryVersionInfo()
+void SourceSocketPoller::ResumePolling(wxSOCKET_T WXUNUSED(sock))
 {
-    const curl_version_info_data* vi = curl_version_info(CURLVERSION_NOW);
-    wxString desc = wxString::Format("libcurl/%s", vi->version);
-    if (vi->ssl_version[0])
-        desc += " " + wxString(vi->ssl_version);
-    return wxVersionInfo("libcurl",
-        vi->version_num >> 16 & 0xff,
-        vi->version_num >> 8 & 0xff,
-        vi->version_num & 0xff,
-        desc);
+}
+
+void SourceSocketPoller::CleanUpSocketSource(wxEventLoopSource* source)
+{
+    wxEventLoopSourceHandler* srcHandler = source->GetHandler();
+    delete source;
+    delete srcHandler;
+}
+
+SocketPollerImpl* SocketPollerImpl::Create(wxEvtHandler* hndlr)
+{
+    return new SourceSocketPoller(hndlr);
+}
+
+#else
```

Comment:
Is it worth having this branch? It seems to me that it's never going to be actually used, is it?

Reply:
I removed the select based socket poller in 4c8f9ff.  I didn't know if there were any systems that wouldn't be covered by the first two cases.